### DISCUSSION TITLE
Additional 0 value in par0 turtle example

### DISCUSSION THREAD

**Discussion by willpkay** [bmcclintock/momentuHMM#74] [Q&A] [answered]
Hi Brett, Théo,

I am looking at the turtle example and I am confused by one of the starting values that you specify for the state-dependent probability distribution parameters. 

In your code you specify the following:

<pre>
nbStates<-2
dist<-list(step="gamma",angle="wrpcauchy")
estAngleMean<-list(angle=TRUE)
circularAngleMean<-list(angle=TRUE)

DM<-list(step=list(mean=~state2(w:angle_osc),sd=~1),
         angle=list(mean=~state2(d),concentration=~1))

Par0<-list(step=c(9.132130, 9.200282, <b>0</b>, 8.579123, 8.640819),
           angle=c(0, -16.0417715, -0.4749668))

turtleFits<-MIfitHMM(miTurtleData$miData, 
                     ncores=ncores,       
                     nbStates=nbStates,
                     dist=dist,
                     Par0=Par0,
                     DM=DM,
                     estAngleMean=estAngleMean,
                     circularAngleMean=circularAngleMean,
                     retryFits=retryFits)
</pre>

The value that I am confused by is the third value in the `Par0` object (the zero; in bold above, and below). What is this zero for? Is it a zero-mass parameter? 

I thought that in a 2-state HMM, using gamma distribution for step length, you would specify a total of 4 step length parameters (2 means, and 2 SDs). Here however, you have specified 5?

<pre>
> Par0$step
[1] 9.132130 9.200282 <b>0.000000</b> 8.579123 8.640819
</pre>

I am trying to repeat your analysis using my own data, and I am using three states: Resting, Foraging, and Transiting. I would like, as you have done, to model the transit state as being potentially influenced by ocean surface currents, and the other two states to be unaffected. I have hence tried to mirror your code using three states, as follows:

```
nbStates<-3
dist<-list(step="gamma",angle="wrpcauchy") 
estAngleMean<-list(angle=TRUE)
circularAngleMean<-list(angle=TRUE)

Par0<-list(step=c(0.89,0.64,0.30,0.13,0.15,0.19),
           angle=c(0,0,0,20,2.31,0.93))

DM<-list(step=list(mean=~state1(tideVel:angle_osc),sd=~1),
           angle=list(mean=~state1(tideDir2),concentration=~1))

Fits <- fitHMM(mHMM,   
               nbStates = nbStates,
               dist = dist,
               Par0 = Par0,
               DM = DM,
               estAngleMean = estAngleMean,
               circularAngleMean = circularAngleMean,
               retryFits = retryFits)
```

Note that in my example the first set of parameter estimates correspond to the transit state, hence I am using `~state1()` as opposed to `~state2()`. Also, my data do not require multiple imputations so I am using `fitHMM()` as opposed to `MIfitHMM()`.

When I try to fit this model however, I get the following error:

```
Error in getDM(data, inputs$DM, inputs$dist, nbStates, p$parNames, p$bounds,  : 
  Dimension mismatch between Par and DM for: step, angle
```

If I have understood the error correctly, it arises because I have not provided enough parameters in the `Par0` argument. Am missing a parameter like the additional **0** that you use above? 

The problem is, I do not understand what the **0** is for and hence I do not know if I should be using zero also, or a different value. I also do not know where I should specify the additional parameter in my `Par0` object because i) my model has 3 states whereas yours has 2 and ii) my starting values are in the sequence transit:forage:rest, whereas yours are forage:transit.

Thanks!

William Kay

**Comment by bmcclintock** [comment, marked as answer]
Hi William,

You are using the design matrix ('DM'), so the 'step' and 'angle' parameters must be provided on the working scale, and the number of parameters is determined by the number of columns in the design matrix. For more on how 'DM' works, see the momentuHMM vignette and help files (e.g. `?fitHMM`). 

A function that may help you figure out the number and order of parameters is 'checkPar0':

`checkPar0(mHMM, nbStates = nbStates, dist = dist, DM = DM, estAngleMean = estAngleMean, circularAngleMean = circularAngleMean)`

Cheers,
Brett

**Comment by willpkay** [comment]
Brilliant! Thank you Brett, especially for such swift advice. I have read the vignette in detail, including several other examples of yours, but was still having a bit of difficulty. Your advice has helped me make progress now!

`checkPar0()` was perfect for determining the number and order of parameters. And I used `getParDM` to convert the parameters to their working scale equivalents. It didn't help that I had previously been playing with the von Mises distribution and hence my concentration parameters were all >0 rather than between 0 and 1. I had forgotten to convert them initially, but they should be correct now..

```
nbStat> nbStates <- 3
> dist <- list(step = "gamma", 
+              angle = "wrpcauchy") 
> estAngleMean <- list(angle = TRUE)
> circularAngleMean <- list(angle = TRUE)
> DM <- list(step = list(mean = ~ state1(tideVel:angle_osc), sd = ~ 1),
+            angle = list(mean = ~ state1(tideDir2), concentration = ~ 1))
> 
> checkPar0(mHMM, 
+           nbStates = nbStates, 
+           dist = dist, 
+           DM = DM, 
+           estAngleMean = estAngleMean, 
+           circularAngleMean = circularAngleMean)

Regression coeffs for step parameters:
--------------------------------------
     mean_1:(Intercept) mean_1:tideVel:angle_osc mean_2:(Intercept) mean_3:(Intercept) sd_1:(Intercept)
[1,]                  1                        2                  3                  4                5
     sd_2:(Intercept) sd_3:(Intercept)
[1,]                6                7

Regression coeffs for angle parameters:
---------------------------------------
     mean_1:(tideDir2) concentration_1:(Intercept) concentration_2:(Intercept) concentration_3:(Intercept)
[1,]                 1                           2                           3                           4

Regression coeffs for the transition probabilities:
---------------------------------------------------
            1 -> 2 1 -> 3 2 -> 1 2 -> 3 3 -> 1 3 -> 2
(Intercept)      1      2      3      4      5      6

Initial distribution:
---------------------
state 1 state 2 state 3 
      1       2       3 
> 
> getPar <- list(step =  c(0.89, 0.64, 0.30, 0.13, 0.15, 0.19), 
+                angle = c(0, 0, 0, 0.6, 0.1, 0.05))
> 
> getParDM(data = mHMM,
+          nbStates = nbStates,
+          dist = dist,
+          Par = getPar,
+          estAngleMean = estAngleMean,
+          circularAngleMean = circularAngleMean,
+          DM = DM)
$`step`
[1] -0.11477590  0.01420444 -0.44628710 -1.20397280 -2.04022083 -1.89711998 -1.66073121

$angle
[1]  1.155193e-09  4.054651e-01 -2.197225e+00 -2.944439e+00

> 
> Par0 <- list(step =  c(-0.11477590, 0.01420444, -0.44628710, -1.20397280, -2.04022083, -1.89711998, -1.66073121), 
+              angle = c(1.155193e-09, 4.054651e-01, -2.197225e+00, -2.944439e+00))
> 
> retryFits <- 10
> 
> Fits <- fitHMM(mHMM,
+                nbStates = nbStates,
+                dist = dist,
+                Par0 = Par0,
+                DM = DM,
+                estAngleMean = estAngleMean,
+                circularAngleMean = circularAngleMean,
+                retryFits = retryFits)
```

And the model has converged..

```
=======================================================================
Fitting HMM with 3 states and 2 data streams
-----------------------------------------------------------------------

 step ~ gamma(mean=~state1(tideVel:angle_osc), sd=~1)
 angle ~ wrpcauchy(mean=~state1(tideDir2), concentration=~1)

 Transition probability matrix formula: ~1

 Initial distribution formula: ~1
=======================================================================
```

Thanks again!

Will
